pkg/controlplane/controller.go

```diff
@@ -174,47 +174,27 @@ func (c *Controller) Start() {
 	repairClusterIPs := servicecontroller.NewRepair(c.ServiceClusterIPInterval, c.ServiceClient, c.EventClient, &c.ServiceClusterIPRange, c.ServiceClusterIPRegistry, &c.SecondaryServiceClusterIPRange, c.SecondaryServiceClusterIPRegistry)
 	repairNodePorts := portallocatorcontroller.NewRepair(c.ServiceNodePortInterval, c.ServiceClient, c.EventClient, c.ServiceNodePortRange, c.ServiceNodePortRegistry)
 
-	// We start both repairClusterIPs and repairNodePorts to catch events
-	// coming from the RunOnce() calls to them.
-	// TODO: Refactor both controllers to only expose a public RunUntil
-	// method, but to accommodate the usecase below (of forcing the first
-	// successful call before proceeding) let them signal on that, like:
-	//   func (c *Repair) RunUntil(stopCh chan struct{}, onFirstSuccess func())
-	// and use it here like:
-	//   wg := sync.WaitGroup{}
-	//   wg.Add(2)
-	//   runRepairClusterIPs := func(stopCh chan struct{}) {
-	//     repairClusterIPs(stopCh, wg.Done)
-	//   }
-	//   runRepairNodePorts := func(stopCh chan struct{}) {
-	//     repairNodePorts(stopCh, wg.Done)
-	//   }
-	//   c.runner = ...
-	//   c.runner.Start()
-	//   wg.Wait()
-	c.stopCh = make(chan struct{})
-	repairClusterIPs.Start(c.stopCh)
-	repairNodePorts.Start(c.stopCh)
-
-	// run all of the controllers once prior to returning from Start.
-	if err := repairClusterIPs.RunOnce(); err != nil {
-		// If we fail to repair cluster IPs apiserver is useless. We should restart and retry.
-		klog.Fatalf("Unable to perform initial IP allocation check: %v", err)
```

Comment:
we no longer fail if we don't allocate?

Reply:
Not explicitly, but we do implicitly.

So what happens, is that this Start (thanks to line 191: `wg.Wait()`) is synchronous in a sense that it waits until first attempt suceeds.
OTOH, Start() is called in pre-start-hook:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/controller.go#L149

So if we won't succeed, pre-start-hook doesn't succeed, so the cluster won't pass livez/ or readyz/ so in generall will get restarted by kubelet.

Does that make sense?

Reply:
yeah, but that assumes a containerized apiserver, a standalone binary will keep running ... I don't know if we should conservative, people using systemd per example for managing apiservers will face a different behavior?

Reply:
I would repeat what I wrote above:

> I personally feel that increasing the interface for the reason of somewhat artificial backward compatibility isn't worth it. I don't think we ever clearly document that we will fatal on if we can't repair even once at start.

But it's actually useful place to get a third voice.

@thockin - WDYT?

Reply:
and if we meet in the middle :smile: 
```
    done := make(chan struct{})
    go func() {
        defer close(done)
        wg.Wait()
    }()
    select {
    case <-done:
    case <-time.After(60*time.Seconds):
        klog.Fatalf("Unable to perform initial IP and Port allocation check: %v", err)
    }
```

Reply:
I like it - changed :)